jax/_src/interpreters/pxla.py

```diff
@@ -3225,21 +3225,49 @@ def create_cpp_call(self, params: stages.CompiledCallParams):
       return None
 
     def aot_cache_miss(*args, **kwargs):
-      # args do not include the const args
+      # args do not include the const args.
+      # We fix the const args to account for 64-bit const args when the
+      # lowering and execution use different `enable_x64` mode.
       # See https://docs.jax.dev/en/latest/internals/constants.html.
-      outs, out_flat, args_flat = stages.Compiled.call(params, *args, **kwargs)
+      fixed_const_args = []
+      for c, c_arg_info, c_shard, c_layout in \
+          zip(params.const_args,
+              self._all_args_info.in_avals[:len(params.const_args)],
+              self._in_shardings[:len(params.const_args)],
+              self._dispatch_in_layouts[:len(params.const_args)]):
+
+        can_c = xla.canonicalize_dtype(c)
+        # can_c does not match the dtype for which the code was compiled
+        if can_c.dtype != c_arg_info.dtype:
+          if (can_c.dtype.itemsize == 8 and
+              dtypes._dtype_to_32bit_dtype.get(can_c.dtype) == c_arg_info.dtype):
+            # 64-bit constant, with 64-bit execution, but 32-bit lowering.
+            with config.enable_x64(False):
```

Comment:
I wonder if there is a better way of doing this, without using `with config.enable_x64`. I need to put the constant on the device either as is (64-bit) even if we are now in 32-bit mode, or as a 32-bit value if the compilation was in 32-bit mode.